src/sys/statfs.rs

```diff
@@ -1,22 +1,120 @@
-use {Result, NixPath};
+use {NixPath, Result};
 use errno::Errno;
+use libc::{self, c_ulong};
+use std::mem;
 use std::os::unix::io::AsRawFd;
-use libc;
 
-pub fn statfs<P: ?Sized + NixPath>(path: &P, stat: &mut libc::statfs) -> Result<()> {
-    unsafe {
-        Errno::clear();
-        let res = try!(
-            path.with_nix_path(|path| libc::statfs(path.as_ptr(), stat))
-        );
+#[cfg(target_os = "android")]
+pub type fsid_t = libc::__fsid_t;
+#[cfg(not(target_os = "android"))]
+use libc::fsid_t;
+
+#[derive(Clone, Copy)]
+#[allow(missing_debug_implementations)]
+pub struct Statfs(libc::statfs);
+
+impl Statfs {
+    /// Magic code defining system type
+    pub fn filesystem_type(&self) -> c_ulong {
+        self.0.f_type as c_ulong
+    }
+
+    /// Optimal transfer block size
+    #[cfg(any(target_os = "freebsd", target_os="macos", target_os="ios"))]
+    pub fn optimal_transfer_size(&self) -> c_ulong {
+        self.0.f_iosize as c_ulong
+    }
+
+    /// Optimal transfer block size
+    #[cfg(not(any(target_os = "freebsd", target_os="macos", target_os="ios")))]
+    pub fn optimal_transfer_size(&self) -> c_ulong {
+        self.0.f_bsize as c_ulong
```

Comment:
This API is almost as bad as `f_type`.  It is variously defined as `uint64_t`, `c_uint`, `c_ulong`, or `c_long`.  In cases like this, Nix usually exposes the underlying feature with as little modification as possible, except for what's required to make it Rust-friendly.  In this case, I think Nix should use the native type of each platform instead of casting everything to the lowest common denominator.